### PR TITLE
Issue #2507: Make Contact Types All Checked By Default on Edit Case Details Page (update)

### LINKED ISSUE #2507: Make Contact Types All Checked By Default on Edit Case Details Page
**What type(s) of user does this feature affect?**
 - supervisors
 - admins

**Description**
Currently, on the Edit Case Details page, all the boxes under contact types are unchecked by default. We would like all of these boxes to be checked by default instead. Also we would like to add two buttons, one which checks all the boxes, and one which unchecks all of the boxes. Finally, we would like all the boxes being unchecked to prevent the page from being saved, with a message pop up that says, "At least one contact type must be selected.".

There is another ticket relating to the layout of this page, which may be relevant: https://github.com/rubyforgood/casa/issues/2505

**Screenshots of current behavior, if any**  
Contact Types in Edit Case Details

_QA Login Details:_  
[Link to QA site](https://casa-qa.herokuapp.com/)  

Login Emails: 
- volunteer1@example.com  view site as a volunteer
- supervisor1@example.com view site as a supervisor
- casa_admin1@example.com view site as an admin

password for all users: 12345678

### PR DESCRIPTION
**Pull request by mussajoop**
### What github issue is this PR for, if any?
Resolves #2507 

### What changed, and why?
- Make Contact Types All Checked By Default on Edit Case Details Page
- Add `check all` and `uncheck all` buttons
- Prevent update action when no contact type has been checked, show the message: _At least one contact type must be selected_
- Add model validation to validate the presence of at least one case contact type on update
### How will this affect user permissions?
- Volunteer permissions: n/a
- Supervisor permissions: n/a
- Admin permissions: n/a

### How is this tested? (please write tests!) 💖💪
- System test added for: check all and uncheck all behavior

### Screenshots please :)
image

image

### CODE REVIEW

File: app/models/casa_case.rb [hunk 1]
```diff
@@ -36,7 +38,7 @@ class CasaCase < ApplicationRecord
   belongs_to :judge, optional: true
   belongs_to :casa_org
   validates :birth_month_year_youth, presence: true
-
+  validates_presence_of :casa_case_contact_types, message: ": At least one contact type must be selected", if: :validate_contact_type
```
Comment (compwron) [inline app/models/casa_case.rb:41]: interesting phrasing here `": At` for error composition I assume

File: app/controllers/casa_cases_controller.rb [hunk 2]
```diff
@@ -63,7 +63,7 @@ def create
   def update
     authorize @casa_case
     original_attributes = @casa_case.full_attributes_hash
-
+    @casa_case.validate_contact_type = true unless current_role == "Volunteer"
```
Comment (compwron) [inline app/controllers/casa_cases_controller.rb:66]: the rest of the time it's false? this feels odd but if it works we will go with it :)
maybe we could make a declarative method name about this

File: app/views/case_contacts/_contact_types.html.erb [hunk 3]
```diff
@@ -35,3 +48,19 @@
     </tr>
   </table>
 </div>
+
+<script type="text/javascript">
```
Comment (compwron) [inline app/views/case_contacts/_contact_types.html.erb:52]: let's test our js more! :)

File: spec/system/casa_cases/edit_spec.rb [hunk 4]
```diff
@@ -408,6 +408,66 @@ def sign_in_and_assign_volunteer
...
+      visit edit_casa_case_path(casa_case)
+      expect(page).to have_text("Select All Contact Types")
+      find("a#uncheck_all").click
+      expect(all("input[type=checkbox][class~=case-contact-contact-type]")).not_to be_empty
+
+      within ".actions" do
+        click_on "Update CASA Case"
+      end
+
+      expect(page).to have_text("At least one contact type must be selected")
```
Comment (compwron) [inline spec/system/casa_cases/edit_spec.rb:469]: nice tests :)

**Review by compwron — APPROVED** [review]

**Comment by compwron** [comment]
woot!!

File: app/models/casa_case.rb (hunk 1 above)
Comment (mussajoop) [reply app/models/casa_case.rb:41]: Yeah its for error composition to have something like '_attribute_name : error_message_'

File: app/controllers/casa_cases_controller.rb (hunk 2 above)
Comment (mussajoop) [reply app/controllers/casa_cases_controller.rb:66]: yeah it's a good idea

File: app/views/case_contacts/_contact_types.html.erb (hunk 3 above)
Comment (mussajoop) [reply app/views/case_contacts/_contact_types.html.erb:52]: woups it passed my 'test radars' 🤖

**Pull request merged**